### ISSUE TITLE
token_mult_prob_error continuously increases during GRPO training with DeepSeek-R1-0528-Qwen3-8B using Megatron backend, leading to training crash

### ISSUE THREAD

**Issue by binhu-nv** [NVIDIA-NeMo/RL#1801]
Describe the bug
When training DeepSeek-R1-0528-Qwen3-8B model using GRPO with Megatron backend (TP=4), the token_mult_prob_error metric continuously increases as training progresses, eventually causing the training to crash. This indicates a growing discrepancy between the inference engine (vLLM) and the training engine (Megatron) when processing the same samples.
According to the documentation, token_mult_prob_error should stay close to 1.0, and values exceeding 1.05 indicate potential framework issues. In my case, this metric keeps growing throughout training until the run becomes unstable and crashes.
Note: This is a metric degradation issue observed via WandB monitoring, not a runtime error log.
Steps/Code to reproduce bug
Quick reproduction from checkpoint:
You can reproduce this issue by resuming training from the following checkpoint, which will quickly show the increasing token_mult_prob_error:
Checkpoint path: /lustre/fsw/portfolios/coreai/users/bihu/RL/results/grpo-dapo-deepseek-r1-8B-8n64g-megatron-tp4-lr1e-6/step_450
Full reproduction paths:
Project directory: /lustre/fsw/portfolios/coreai/users/bihu/RL
Launch script: /lustre/fsw/portfolios/coreai/users/bihu/RL/run_grpo_dapo_deepseek_r1_8B.sh
Config file: /lustre/fsw/portfolios/coreai/users/bihu/RL/examples/configs/grpo_dapo_deepseek_r1_8B_8n64g_megatron.yaml.bak_no_fp32logits
Configuration file content (grpo_dapo_deepseek_r1_8B_8n64g_megatron.yaml.bak_no_fp32logits):
defaults: "grpo_math_1B.yaml"grpo:  num_prompts_per_step: 32  num_generations_per_prompt: 8loss_fn:  reference_policy_kl_penalty: 0.0  ratio_clip_min: 0.2  ratio_clip_max: 0.28  use_importance_sampling_correction: truepolicy:  model_name: "deepseek-ai/DeepSeek-R1-0528-Qwen3-8B"  train_global_batch_size: 64  train_micro_batch_size: 1  logprob_batch_size: 1  max_total_sequence_length: 16384  dtensor_cfg:    enabled: false  megatron_cfg:    enabled: true    converter_type: "Qwen3ForCausalLM"    tensor_model_parallel_size: 4    pipeline_model_parallel_size: 1    context_parallel_size: 1    sequence_parallel: false    activation_checkpointing: true    optimizer:      optimizer: "adam"      lr: 1.0e-6      min_lr: 1.0e-6      weight_decay: 0.01      bf16: true      fp16: false      params_dtype: "float32"      adam_beta1: 0.9      adam_beta2: 0.999      adam_eps: 1e-8      use_distributed_optimizer: true      use_precision_aware_optimizer: false    scheduler:      lr_decay_style: "constant"  sequence_packing:    enabled: true  generation:    max_new_tokens: 15360    vllm_cfg:      gpu_memory_utilization: 0.85data:  max_input_seq_length: 1024  prompt_file: "examples/prompts/cot.txt"  dataset_name: "DAPOMath17K"cluster:  gpus_per_node: 8  num_nodes: 8
Launch command:
cd /lustre/fsw/portfolios/coreai/users/bihu/RLbash run_grpo_dapo_deepseek_r1_8B.sh
Or directly with uv:
uv run examples/run_grpo_math.py \  --config=/lustre/fsw/portfolios/coreai/users/bihu/RL/examples/configs/grpo_dapo_deepseek_r1_8B_8n64g_megatron.yaml.bak_no_fp32logits \  cluster.num_nodes=8 \  cluster.gpus_per_node=8
Environment:
8 nodes × 8 GPUs = 64 GPUs
NeMo-RL version: r0.5.0 (container: nemo-rl:r0.5.0-e883ac42)
Model: deepseek-ai/DeepSeek-R1-0528-Qwen3-8B
Dataset: BytedTsinghua-SIA/DAPO-Math-17k (via DAPOMath17K loader)
Expected behavior
The token_mult_prob_error metric should remain stable around 1.0 (or at least below 1.05) throughout training, indicating consistent alignment between the vLLM inference engine and the Megatron training engine. Training should complete without crashing due to numerical instability.
Additional context
WandB metrics screenshot:

<img width="1096" height="660" alt="Image" src="https://github.com/user-attachments/assets/411a2131-f569-41d5-999b-df884145ca47" />

<img width="361" height="716" alt="Image" src="https://github.com/user-attachments/assets/e5b1cf13-f736-4afa-9263-8909345ff7ea" />

Observations:
The token_mult_prob_error starts near 1.0 but gradually increases over hundreds of training steps
Training eventually becomes unstable and crashes
Configuration notes:
Using Megatron backend (not DTensor)
TP=4 with converter_type: "Qwen3ForCausalLM"
params_dtype: "float32" for optimizer parameters
No KL penalty (reference_policy_kl_penalty: 0.0)
Long sequence length (max 16k tokens)

**Comment by Hameed92** [comment]
We are facing the same issue here with the standard deepscaler experiment (examples/configs/recipes/llm/grpo-deepscaler-1.5b-8K.yaml). We're using v0.5.0.
When we use dtensor, `token_mult_prob_error` stays around 1. But when we use megatron backend (without changing anything else in the experiment configuration), we get these very large values. 

<img width="731" height="725" alt="Image" src="https://github.com/user-attachments/assets/22ce5c17-1d38-415c-bb2d-c13a6ecc8a50" />

The intriguing thing is that even though  `token_mult_prob_error` is quite wrong, the output is good, and similar to what is reported here `docs/guides/grpo-deepscaler.md`.

**Comment by jinglinglingling** [comment]
```yaml
policy:
  megatron_cfg:
    apply_rope_fusion: false
    bias_activation_fusion: false
```

I bisected this on DeepSeek-R1-Distill-Qwen-1.5B / 8K / GRPO (1n × 8× A100, 109 steps) and found that the config above removes ~88% of `token_mult_prob_error` spikes, bringing Megatron's behavior back close to the DTensor baseline, with no impact on reward.

A "**spike**" here is a step where **`token_mult_prob_error` (tmpe) > 1.05** — the same threshold the DTensor recipe asserts on. Comparison:

| Config | step spike rate | tmpe median | tmpe max | reward |
| --- | --- | --- | --- | --- |
| DTensor (baseline) | 0% | 1.0121 | 1.01 | ~0.5 |
| Megatron default (all fusions ON) | 100% | 38565 | 6.55e+09 | ~0.5 |
| Megatron + `apply_rope_fusion=false` | 24.6% | 520 | 1.36+06 | ~0.5 |
| Megatron + `apply_rope_fusion=false` + `bias_activation_fusion=false` | 6.4% | 1.0116 | 709 | ~0.5 |

**Why**: both `apply_rope_fusion` (rotary position embedding) and `bias_activation_fusion` (bias + SwiGLU) are fused CUDA kernels that keep intermediates in bf16, while vLLM's corresponding native PyTorch ops upcast intermediates to fp32. The per-op rounding gap (~0.4% in bf16 vs ~1e-8 in fp32) is tiny on most tokens, but as training progresses and the policy gets sharper (top-1 / top-2 logits get closer), the two engines occasionally disagree on the top-1 ranking. A single rank flip blows that token's `mult_prob_error` up to `e^15`-scale and dominates the per-batch max. Disabling both fusions routes Megatron through the same fp32-intermediate path as vLLM, so the rounding cancels at training time. Reward is unaffected because GRPO's ratio clip already absorbs these outliers — > 99% of tokens are fine, and the per-batch mean stays at ~1.01.

**Scope**: directly verified on DeepSeek-R1-Distill-Qwen-1.5B (Qwen2 arch, no QK-Norm). The mechanism is generic to any model using these fused kernels (Llama, Qwen2/3, DeepSeek, Mistral, …) on PPO/GRPO with vLLM generation. Models with QK-Norm (Qwen3, Qwen3-MoE) likely see this **amplified** rather than reduced, since QK-Norm squeezes Q/K into a smaller range where the same bf16 absolute error becomes a larger relative error.

**Remaining ~12%**: small spikes (mostly 1.05–1.5) caused by other bf16 divergences (TE FusedAttention vs FlashAttention-2). They don't affect reward, and the assertion threshold can be safely relaxed.

**Scope**: verified on DeepScaler 1.5B; architecturally expected to hold for other QK-Norm + RoPE models (Qwen2/3, DeepSeek-V2/V3). Throughput cost ~5–10%.


<img width="2472" height="316" alt="Image" src="https://github.com/user-attachments/assets/474c325b-9f85-43a3-ac89-2ff3b89ca961" />

**Comment by binhu-nv** [comment]
Great findings, does all the model which has qk norm shouldn't use apply_rope_fusion and  bias_activation_fusion? not only qwen model?

**Comment by jinglinglingling** [comment]
Likely yes for all Megatron + vLLM PPO/GRPO setups, but we only directly verified on DeepSeek-R1-Distill-Qwen-1.5B (Qwen2 arch, no QK-Norm). The fix is safe to apply to any model — worst case it costs ~5-10% throughput. We expect QK-Norm models (Qwen3/MoE) to be equally or more affected, not less.
